src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -708,5 +709,195 @@ public static nuint RotateRight(nuint value, int offset)
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
+        /// * uint32_t __crc32ch (uint32_t a, uint16_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, ushort data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint16_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
+            return Crc32Fallback.Crc32C(crc, data);
+        }
+
+        /// <summary>
+        /// Accumulates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+        /// * uint32_t __crc32ch (uint32_t a, uint32_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, uint data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint32_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
+            return Crc32Fallback.Crc32C(crc, data);
+        }
+
+        /// <summary>
+        /// Accumulates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned __int64 _mm_crc32_u64 (unsigned __int64 crc, unsigned __int64 v)
+        /// * uint32_t __crc32cd (uint32_t a, uint64_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, ulong data)
+        {
+            if (Sse42.X64.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                ulong result = Sse42.X64.Crc32(crc, data);
+
+                return unchecked((uint)result);
+            }
```

Comment:
```suggestion
            if (Sse42.X64.IsSupported)
            {
                return (uint)Sse42.X64.Crc32(crc, data);
            }
            if (Sse42.IsSupported)
            {
                return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), (uint)(data >> 32));
            }
```

This is what I meant about breaking down the `ulong` overload for x86.  Use the 64-bit instruction when possible, but rather than falling back to the software implementation, just break it into halves and call the `uint` intrinsic if it's available.

Same can be done for ARM, except SSE4.2 implies little endian whereas ARM is technically bi-endian so you must byte swap BE input if you want to process in LE order to match the above.